### ISSUE TITLE
Tests using react-dnd with useDrag and useDrop

### ISSUE THREAD

**Issue by jodevan** [react-dnd/react-dnd#1540]
Hello there!

I'm wondering if there's a way to test a component using react-dnd using useDrag and useDrop hooks. 

I posted a question yesterday in StackOverflow that describes my problem: https://stackoverflow.com/questions/57994630/tests-using-react-dnd-with-usedrag-and-usedrop

I also asked for help in https://www.reactiflux.com/ and I was told to open an issue here.

Here's the summary of my question:

Has anyone ever been able to test the drag-and-drop functionality using a functional component with useDrag and useDrop hooks?

According to the examples found here http://react-dnd.github.io/react-dnd/docs/testing, they either decorate the original component with a DragSource or a DropTarget HOC.

I couldn't find any example of tests using their hooks though. They do have examples of code using both decorators and hooks (https://github.com/react-dnd/react-dnd/tree/master/packages/documentation) but there are examples of tests only using the decorators.

Any ideas? 

Thanks!

**Comment by darthtrevino** [comment]
To be honest, that's an area that needs to be addressed, and it has not been so far.

**Comment by netochaves** [comment]
I'm having the same problem, I can't test my functional components using ùseDrop` and `useDrag`. Any workaround or update on this?

**Comment by brapifra** [comment]
It would also be good to have some tutorial about how to test them with `react-testing-library`. I was only able to test the `beginDrag` event:
```typescript
expect(component.queryByText('Drag me')).toBeTruthy();
fireEvent.dragStart(component.getByText('Drag me'));
expect(onDragBegin).toBeCalledTimes(1);
```

**Comment by apflieger** [comment]
A solution is to mock useDrag/useDrop functions. I 'inject' these functions in components props so I can pass a mock in my tests. I don't write tests on drag/drop logic but at least I can make the reste of the test pass.

Here's the code of these mocks https://gist.github.com/apflieger/608816b15fbbf21fae298d2fc888666f

**Comment by brapifra** [comment]
Ideally, you shouldn't need to mock anything :sweat_smile: @apflieger

**Comment by tarjei** [comment]
Hi, any movement on this? I'm curious on what is needed to make this happen. Is the problem that there are no way to query/find the draggable/droppable item to trigger the events? 
T

**Comment by saifbechan** [comment]
Currently using mocks for this also. Would love to see a proper solution for this.

**Comment by vovkvlad** [comment]
I'm stuck on this also. Would be nice to have some movement on this

**Comment by hoangnm** [comment]
I'm also using react-testing-library, just find the correct drag and drop elements, then trigger the correct events, it works:
`
fireEvent.dragStart(screen.getByText('Drag me'));
fireEvent.drop(screen.getByText('Drop here'));
`

**Comment by brapifra** [comment]
Yeah, I'm also using react-testing-library and it works like a charm. No mocking required :slightly_smiling_face: 
```typescript
import { fireEvent } from '@testing-library/react';
...
function dragAndDrop(src: Element, dst: Element) {
    fireEvent.dragStart(src);
    fireEvent.dragEnter(dst);
    fireEvent.drop(dst);
    fireEvent.dragLeave(dst);
    fireEvent.dragEnd(src);
}
```

**Comment by doctorboost** [comment]
> Yeah, I'm also using react-testing-library and it works like a charm. No mocking required 🙂
> 
> ```ts
> import { fireEvent } from '@testing-library/react';
> ...
> function dragAndDrop(src: Element, dst: Element) {
>     fireEvent.dragStart(src);
>     fireEvent.dragEnter(dst);
>     fireEvent.drop(dst);
>     fireEvent.dragLeave(dst);
>     fireEvent.dragEnd(src);
> }
> ```

Hi @brapifra , What are you using for the src and dst? I'm assuming the src is the element you want to move but what about the dst? Is that the list? Thanks

**Comment by brapifra** [comment]
Hey @doctorboost,
`src` is the element that you want to drag, and `dst` is the element that you want drop it into. For example:
```typescript
import { fireEvent, render } from '@testing-library/react';
...
const component = render(<Component ... />);

dragAndDrop(
   component.getByText('drag me'),
   component.getByText('drop here'),
);
```

**Comment by carpeliam** [comment]
Just so I'm clear, the `DecoratedComponent` testing strategy in the [testing documentation](https://react-dnd.github.io/react-dnd/docs/testing) is not designed to work with `useDrag` and `useDrop` which are mentioned in the [tutorial](https://react-dnd.github.io/react-dnd/docs/tutorial), right? If that's the case, given that the testing doc comes immediately after the tutorial doc, I find it more than a little frustrating that the testing doc seems to omit this...

**Comment by amorriscode** [comment]
> Yeah, I'm also using react-testing-library and it works like a charm. No mocking required 🙂
> 
> ```ts
> import { fireEvent } from '@testing-library/react';
> ...
> function dragAndDrop(src: Element, dst: Element) {
>     fireEvent.dragStart(src);
>     fireEvent.dragEnter(dst);
>     fireEvent.drop(dst);
>     fireEvent.dragLeave(dst);
>     fireEvent.dragEnd(src);
> }
> ```

@brapifra this is probably too much to ask but I'm wondering if you could help produce a working example of this with CodeSandbox. I can't get it to work but I'm assuming it's related to how I'm using `useDragLayer` and `useDrag`.

I don't actually have a drop zone, I'm just dragging an element around the screen. Ideally, my test would be:
1. Click and drag component from the bottom right of the screen to any other position
2. Test that the positioning of the element has changed

**Comment by svetozar12** [comment]
the easiest way to mock useDrop and useDrag is like this

![image](https://user-images.githubusercontent.com/76191412/183600993-2413666c-b1a5-4932-9a60-26270998bfd0.png)

**Comment by zapalagrzegorz** [comment]
Would just add my experience. For unknown reason simple 
`fireEvent.dragStart(src);` didn't cause my `isDragging` to change. 
Only when was `fireEvent.dragEnter(second);` added, property ìsOver` turned to be true.

**Comment by tyro-e** [comment]
@svetozar12 this works for me!

**Comment by yanghoxom** [comment]
I was also facing the same issue as @zapalagrzegorz 
isDragging return false in testing-library
